File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/managehourview/HourToggleView.java
package com.wiseassblog.samsaradayplanner.ui.managehourview;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import com.wiseassblog.samsaradayplanner.R;
import com.wiseassblog.samsaradayplanner.domain.constants.QUARTER;

public class HourToggleView extends ConstraintLayout {
    private Switch quarterHourIsActive;
    private TextView timeLabel;
    private Spinner selectTask;
    private View bottomBorder;
    private QUARTER quarter;
    private ToggleViewClickHandler callback;

    public void setToggleState(QUARTER quarter, boolean isActive) {
        if (quarter != QUARTER.ZERO) {
            quarterHourIsActive.setChecked(isActive);
            setTimeLabelColor(isActive);
        } else {
            quarterHourIsActive.setVisibility(INVISIBLE);
            setTimeLabelColor(true);
        }

       if (quarter == QUARTER.FOURTY_FIVE) hideBottomDivider();
    }

    private void setTimeLabelColor(boolean isAccentColor) {
        if (isAccentColor) timeLabel.setTextColor(
                ContextCompat.getColor(
                        getContext(),
                        R.color.colorAccent
                )
        );
        else {
            timeLabel.setTextColor(
                    ContextCompat.getColor(
                            getContext(),
                            android.R.color.white
                    )
            );
        }
    }

    interface ToggleViewClickHandler {
        void onQuarterToggle(QUARTER quarter, boolean activate);

        /**
         * @param quarter  the Quarter which represents one of the four HourToggleViews
         * @param position the position of the task which was selected from the ArrayAdapter's data.
         *                 From there we can determine how to update the UI and storage
         */
        void onTaskSelected(QUARTER quarter, int position);
    }


    public HourToggleView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        build();
    }

    private void build() {
        inflate(getContext(), R.layout.view_quarter_hour_toggle, this);

        quarterHourIsActive = findViewById(R.id.swi_activate_quarter_hour);
        quarterHourIsActive.setOnCheckedChangeListener(
                (view, isChecked) -> {
                    if (callback != null) callback.onQuarterToggle(quarter, isChecked);
                    setTimeLabelColor(isChecked);
                }
        );

        timeLabel = findViewById(R.id.lbl_quarter_hour);

        selectTask = findViewById(R.id.spn_quarter_hour);
        bottomBorder = findViewById(R.id.view_bottom_border);
    }

    public void setCallback(ToggleViewClickHandler callback) {
        this.callback = callback;

        selectTask.setOnItemSelectedListener(
                new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> adapterView,
                                               View view,
                                               int position,
                                               long l) {
                        if (callback != null) callback.onTaskSelected(quarter, position);
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> adapterView) {}
                }
        );
    }

    /**
     * @param text Expected to be a properly formatted time display like 1:00PM or 13:00
     */
    public void setTimeLabel(String text) { timeLabel.setText(text);}

    public void hideBottomDivider() { bottomBorder.setVisibility(INVISIBLE); }

    public QUARTER getQuarter() { return quarter;}

    public void setQuarter(QUARTER quarter) { this.quarter = quarter;}

    public void createTaskAdapter(String[] tasks, int startingPosition) {
        ArrayAdapter<String> adapter = new ArrayAdapter(getContext(), android.R.layout.simple_spinner_dropdown_item, tasks);

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        selectTask.setAdapter(adapter);

        selectTask.setSelection(startingPosition, false);
    }
}

File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/managehourview/HourToggleView.kt
package com.wiseassblog.samsaradayplanner.ui.managehourview

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import com.wiseassblog.samsaradayplanner.R
import com.wiseassblog.samsaradayplanner.domain.constants.QUARTER

class HourToggleView(context: Context, attrs: AttributeSet?) : ConstraintLayout(context, attrs) {
    private lateinit var quarterHourIsActive: Switch
    private lateinit var timeLabel: TextView
    private lateinit var selectTask: Spinner
    private lateinit var bottomBorder: View

    lateinit var quarter: QUARTER

    var onQuarterToggle: ((quarter: QUARTER, activate: Boolean) -> Unit)? = null

    fun setToggleState(quarter: QUARTER, isActive: Boolean) {
        if (quarter != QUARTER.ZERO) {
            quarterHourIsActive!!.isChecked = isActive
            setTimeLabelColor(isActive)
        } else {
            quarterHourIsActive!!.visibility = INVISIBLE
            setTimeLabelColor(true)
        }
        if (quarter == QUARTER.FOURTY_FIVE) hideBottomDivider()
    }

    private fun setTimeLabelColor(isAccentColor: Boolean) {
        if (isAccentColor) timeLabel!!.setTextColor(
            ContextCompat.getColor(
                context,
                R.color.colorAccent
            )
        ) else {
            timeLabel!!.setTextColor(
                ContextCompat.getColor(
                    context,
                    android.R.color.white
                )
            )
        }
    }

    private fun build() {
        inflate(context, R.layout.view_quarter_hour_toggle, this)
        quarterHourIsActive = findViewById(R.id.swi_activate_quarter_hour)
        quarterHourIsActive.setOnCheckedChangeListener(
            { view: CompoundButton?, isChecked: Boolean ->
                onQuarterToggle?.let { it(quarter, isChecked) }
                setTimeLabelColor(isChecked)
            }
        )
        timeLabel = findViewById(R.id.lbl_quarter_hour)
        selectTask = findViewById(R.id.spn_quarter_hour)
        bottomBorder = findViewById(R.id.view_bottom_border)
    }

    fun setCallback(
        onQuarterToggle: (quarter: QUARTER, activate: Boolean) -> Unit,
        onTaskSelected: (quarter: QUARTER, position: Int) -> Unit
    ) {
        this.onQuarterToggle = onQuarterToggle


        selectTask.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, position: Int, p3: Long) {
                onTaskSelected.invoke(
                    quarter,
                    position
                )
            }

            override fun onNothingSelected(p0: AdapterView<*>?) {

            }

        }
    }

    /**
     * @param text Expected to be a properly formatted time display like 1:00PM or 13:00
     */
    fun setTimeLabel(text: String?) {
        timeLabel!!.text = text
    }

    fun hideBottomDivider() {
        bottomBorder!!.visibility = INVISIBLE
    }

    fun createTaskAdapter(tasks: Array<String>, startingPosition: Int) {
        val adapter: ArrayAdapter<String> =
            ArrayAdapter<String>(context, android.R.layout.simple_spinner_dropdown_item, tasks)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        selectTask!!.adapter = adapter
        selectTask!!.setSelection(startingPosition, false)
    }

    init {
        build()
    }
}